internal/logging/logging.go

```diff
@@ -0,0 +1,195 @@
+			return zap.FatalLevel.Enabled(lvl) || enabled.Load() && atom.Enabled(lvl)
+		}),
+	))
+
+	return &Logging{
+			enabled: enabled,
+			level:   atom,
+			output:  output,
+			logger:  logger.Sugar(),
+			encoder: encoder,
+			syncer:  syncer,
+			loggers: map[string]*zap.SugaredLogger{},
+			options: options,
+		},
+		nil
+}
+
+// GetChildLogger returns a named child logger.
+// Log levels for named child loggers are obtained from the logging options and, if not found,
+// set to the default log level.
+func (l *Logging) GetChildLogger(name string) *zap.SugaredLogger {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	if logger, ok := l.loggers[name]; ok {
+		return logger
+	}
+
+	if level, found := l.options[name]; found {
+		atom := zap.NewAtomicLevelAt(level)
+
+		logger := l.logger.Desugar().WithOptions(
+			zap.WrapCore(func(c zapcore.Core) zapcore.Core {
+				return zapcore.NewCore(
+					l.encoder,
+					l.syncer,
+					zap.LevelEnablerFunc(func(lvl zapcore.Level) bool {
+						// Always log fatals, but allow other levels to be completely disabled.
+						return zap.FatalLevel.Enabled(lvl) || l.enabled.Load() && atom.Enabled(lvl)
+					}))
+			})).Sugar().Named(name)
+
+		l.loggers[name] = logger
+
+		return logger
+	}
+
+	logger := l.logger.Named(name)
+	l.loggers[name] = logger
+
+	return logger
+}
+
+// GetLogger returns the default logger.
+func (l *Logging) GetLogger() *zap.SugaredLogger {
+	return l.logger
+}
+
+const (
+	ExitFailure = 1
+)
+
+// Fatal logs fatal errors and exits afterwards.
+//
+// Sleeps for a short time before exiting if the output is set to systemd-journal,
+// as systemd may not correctly attribute all or part of the log message due to a race condition
+// and therefore it would be missing when checking logs via journalctl -u icingadb:
+// https://github.com/systemd/systemd/issues/2913
+//
+// Without sleep, we observed the following exit behavior when starting Icinga DB via systemd:
+//
+// • All or part of the log message is lost,
+// regardless of whether it is logged to stderr or to the journal.
+//
+// • When writing to stderr after also writing to the journal,
+// the message seems to be attributed correctly,
+// which is good for panics not under our control as they print to stderr.
+// For messages under our control,
+// this is actually not an option,
+// as we only want to log to the journal.
+//
+// • Sleep before exiting helps systemd to attribute the message correctly,
+// regardless of whether it is written to the journal or stderr.
+func (l *Logging) Fatal(err error) {
+	logger := l.GetLogger().Desugar()
+
+	if l.output == "systemd-journal" {
+		l.enabled.Store(false)
+		// Fatal would exit after write, but we want to sleep first in case of systemd-journal logging.
+		logger = logger.WithOptions(zap.OnFatal(zapcore.WriteThenGoexit))
+	}
+
+	logged := make(chan struct{})
+	go func() {
+		defer close(logged)
+		defer logger.Sync()
+		logger.Fatal(fmt.Sprintf("%+v", err))
+	}()
+	<-logged
```

Comment:
Zap doesn't let you use `WriteThenNoop` in` Fatal`.

Reply:
Another option might be to just sleep in the writer (or encoder should we implement that) right after sending the message to journald. Then zap could just exit as normal.

Reply:
Is there really a benefit since we would have to pass `logging.enabled` to the writer and always check whether we're dealing with the `fatal` level?

Reply:
Apart from getting rid of this `WriteThenGoexit` hack, it would also allow using the regular `logger.Fatal*()` functions that currently wouldn't do the sleep.